hbase-coprocessor/src/main/java/org/apache/omid/transaction/TransactionVisibilityFilter.java

```diff
@@ -0,0 +1,248 @@
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), v.getTimestamp());
+                return ReturnCode.NEXT_COL;
+            }
+
+            if (shadowCellCache.containsKey(v.getTimestamp()) &&
+                    hbaseTransaction.getStartTimestamp() >= shadowCellCache.get(v.getTimestamp())) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), shadowCellCache.get(v.getTimestamp()));
+                return ReturnCode.NEXT_COL;
+            }
+
+            // Try to get shadow cell from region
+            final Get get = new Get(CellUtil.cloneRow(v));
+            get.setTimeStamp(v.getTimestamp()).setMaxVersions(1);
+            get.addColumn(CellUtil.cloneFamily(v), CellUtils.addShadowCellSuffix(CellUtils.FAMILY_DELETE_QUALIFIER));
+            Result deleteFamilySC = snapshotFilter.getTableAccessWrapper().get(get);
+
+            if (!deleteFamilySC.isEmpty() &&
+                    Bytes.toLong(CellUtil.cloneValue(deleteFamilySC.rawCells()[0] )) < hbaseTransaction.getStartTimestamp()){
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), Bytes.toLong(CellUtil.cloneValue(deleteFamilySC.rawCells()[0])));
+                return ReturnCode.NEXT_COL;
+            }
+
+            //At last go to commit table
+            Optional<Long> commitTimestamp = snapshotFilter.tryToLocateCellCommitTimestamp(hbaseTransaction.getEpoch(),
+                    v, shadowCellCache);
+            if (commitTimestamp.isPresent() && hbaseTransaction.getStartTimestamp() >= commitTimestamp.get()) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), commitTimestamp.get());
+                return ReturnCode.NEXT_COL;
+            }
+
+            // Continue getting the next version of the delete family,
+            // until we get one in the snapshot or move to next cell
+            return ReturnCode.SKIP;
+        }
+
+        if (familyDeletionCache.containsKey(Bytes.toString(CellUtil.cloneFamily(v)))
+                && familyDeletionCache.get(Bytes.toString(CellUtil.cloneFamily(v))) >= v.getTimestamp()) {
+            return ReturnCode.NEXT_COL;
+        }
+
+        if (isCellInSnapshot(v)) {
+
+            if (CellUtils.isTombstone(v)) {
+                return ReturnCode.NEXT_COL;
+            }
+
+            if (hbaseTransaction.getVisibilityLevel() == AbstractTransaction.VisibilityLevel.SNAPSHOT) {
+                return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
+            } else if (hbaseTransaction.getVisibilityLevel() == AbstractTransaction.VisibilityLevel.SNAPSHOT_ALL) {
+                if (snapshotFilter.getTSIfInTransaction(v, hbaseTransaction).isPresent()) {
+                    return runUserFilter(v, ReturnCode.INCLUDE);
+                } else {
+                    return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
```

Comment:
I would add a comment that since v is in snapshot and not in transaction then it is the last result for SNAPSHOT_ALL.